main.go

```diff
@@ -311,50 +310,64 @@ type promCache struct {
 	cache *ristretto.Cache
 }
 
-func (p *promCache) Query(ctx context.Context, query string, ts time.Time) (model.Value, prometheusv1.Warnings, error) {
-	xxh := xxhash.New()
-	_, _ = xxh.WriteString(query)
-	hash := xxh.Sum64()
+type promCacheKeyType string
+
+const promCacheKey promCacheKeyType = "promCache"
+
+func contextSetPromCache(ctx context.Context, t time.Duration) context.Context {
+	return context.WithValue(ctx, promCacheKey, t)
+}
 
-	if value, exists := p.cache.Get(hash); exists {
+func contextGetPromCache(ctx context.Context) time.Duration {
+	t, ok := ctx.Value(promCacheKey).(time.Duration)
+	if ok {
+		return t
+	}
+	return 0
+}
+
+func (p *promCache) Query(ctx context.Context, query string, ts time.Time) (model.Value, prometheusv1.Warnings, error) {
+	if value, exists := p.cache.Get(query); exists {
 		return value.(model.Value), nil, nil
 	}
 
+	start := time.Now()
 	value, warnings, err := p.api.Query(ctx, query, ts)
+	duration := time.Since(start)
 	if err != nil {
-		return nil, warnings, fmt.Errorf("prometheus Query: %w", err)
+		return nil, warnings, fmt.Errorf("prometheus query: %w", err)
 	}
 
-	if v, ok := value.(model.Vector); ok {
-		if len(v) > 0 && len(warnings) == 0 {
-			// TODO might need to pass cache duration via ctx?
-			_ = p.cache.SetWithTTL(hash, value, 10, 5*time.Minute)
+	cacheDuration := contextGetPromCache(ctx)
+	if cacheDuration > 0 {
+		if v, ok := value.(model.Vector); ok {
+			if len(v) > 0 && len(warnings) == 0 {
+				_ = p.cache.SetWithTTL(query, value, duration.Milliseconds(), cacheDuration)
+			}
 		}
 	}
 
 	return value, warnings, nil
 }
 
 func (p *promCache) QueryRange(ctx context.Context, query string, r prometheusv1.Range) (model.Value, prometheusv1.Warnings, error) {
-	xxh := xxhash.New()
-	_, _ = xxh.WriteString(query)
-	_, _ = xxh.WriteString(r.Start.String())
-	_, _ = xxh.WriteString(r.End.String())
-	hash := xxh.Sum64()
```

Comment:
Just found out why the graphs aren't properly loading anymore. We cache every query once now no matter what the time range that what requested... 
![errorbudget](https://user-images.githubusercontent.com/872251/159713092-f74b30a6-8544-43d0-9af2-94ef66ec2a88.png)

The reason this only shows half a day is because before I requested this for 12h... :joy: 
We'll need to include the start and end into the caching. Maybe we can somehow defer the duration from that and round it. At least that way we have some better caching. If we were to simply use Start and End like before which are different every second, it'll never actually cache...